### ISSUE TITLE
`?` error propagation lint breaks with try-blocks

### ISSUE THREAD

**Issue by DaniD3v** [rust-lang/rust-clippy#12337]
### Summary

See title.

Specifically the lint for `let Some(var_name) = ... else { return None; }` breaks,
but I suspect the other ? propagation lints might break as well

### Lint Name

clippy::question_mark

### Reproducer

I tried this code:

```rust
fn returns_option() -> Option<()> {
  let _: Option<()> = try {
    let Some(_) = Some(()) else { return None; };
  };
  Some(())
}
```

I saw this happen:

```
warning: this `let...else` may be rewritten with the `?` operator
  --> src/digi4school/book.rs:65:5
   |
65 |     let Some(_) = Some(()) else { return None; };
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace it with: `let _ = Some(())?;`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#question_mark
   = note: `-W clippy::question-mark` implied by `-W clippy::style`
   = help: to override `-W clippy::style` add `#[allow(clippy::question_mark)]`
```

I expected to see this happen:

No lint

### Version

```text
rustc 1.78.0-nightly (3406ada96 2024-02-21)
binary: rustc
commit-hash: 3406ada96f8e16e49e947a91db3eba0db45245fa
commit-date: 2024-02-21
host: x86_64-unknown-linux-gnu
release: 1.78.0-nightly
LLVM version: 18.1.0
```

**Comment by y21** [comment]
There are no try blocks in the reproducer, but I assume what you meant is something like
```rs
fn returns_option() -> Option<()> {
  let _: Option<()> = try {
    let Some(_) = Some(()) else { return None; };
    ()
  };
  Some(())
}
```

@rustbot claim

**Comment by DaniD3v** [comment]
> There are no try blocks in the reproducer, but I assume what you meant is something like
> 
> ```rust
> fn returns_option() -> Option<()> {
>   let _: Option<()> = try {
>     let Some(_) = Some(()) else { return None; };
>   };
>   Some(())
> }
> ```
> 
> @rustbot claim

Yeah sorry totally forgot about the **main point of the issue** when I minified it. :grimacing: 
I edited the original post. Should be correct now
